### ISSUE TITLE
Are globals supported?

### ISSUE THREAD

**Issue by tremby** [justjohn/twig.js#317]
See http://twig.sensiolabs.org/doc/advanced.html#globals

I can't find any code or wiki references.

**Comment by lvl99** [comment]
+1

**Comment by MichaelMackus** [comment]
:+1:

**Comment by JohnAlbin** [comment]
(Making this issue more Google-friendly as I had a hard time finding it.)

The 3 global variables documented at http://twig.sensiolabs.org/doc/templates.html#global-variables are:

> The following variables are always available in templates:
> 
> `_self`: references the current template;
> `_context`: references the current context;
> `_charset`: references the current charset.

Note that http://twig.sensiolabs.org/doc/deprecated.html#globals says:
> As of Twig 2.x, the ability to register a global variable after the runtime or the extensions have been initialized is not possible anymore (but changing the value of an already registered global is possible).
> As of Twig 1.x, using the `_self` global variable to get access to the current `Twig_Template` instance is deprecated; most usages only need the current template name, which will continue to work in Twig 2.0. **In Twig 2.0, `_self` returns the current template name** instead of the current `Twig_Template` instance.

[Highlighting mine.]

Since "most usages only need the current template name", I think we should do the simpler implementation where `_self` returns the name of the template.

**Comment by tremby** [comment]
Those aren't what I meant -- I'd never even heard of those. What I was referring to are user-specified global variables -- variables defined at Twig setup time which are then available to all templates no matter the context sent to the template at render time -- as documented in the link I gave.

**Comment by gsouf** [comment]
Without global I cannot easyly use my php twig templates in twig js. I have to inject globals every times I render a template.

**Comment by lvl99** [comment]
It's been a long time since I last saw this, but you could effectively write a plugin to manage your globals. Here's some dummy code which I haven't tested (i.e. don't copy and paste! Prob won't work first try):
```javascript
// site-globals.js
import objectPath from "object-path"

const siteGlobals = {
  env: "development",
  config: {
    domain: "example.com"
  }
}

Twig.extendFunction("siteGlobal", function(path, defaultValue) {
    return objectPath.get(siteGlobals, path, defaultValue)
})
```
```twig
{# example.html #}

{{ siteGlobal("config.domain", "defaultvalue.com") }}
```

If you use Twig.js with Express, you can also initialise it with whatever variables (aka globals?) when you render the layout (here's an example which uses [`config`](https://www.npmjs.com/package/config) to load in the site's config to use in the Twig template:
```javascript
// config.json
{
  "domain": "example.com"
}
```
```javascript
// index.js
import express from 'express'
import { twig }  from 'twig'
import config from 'config'

const app = express()
app.set('view engine', 'twig')
app.get '/', (req, res) => {
  const page = { title: "Example" }
  res.render('index', {
    // Set the data when the template is rendered as a response
    env: process.env.NODE_ENV,
    config,
    page
  })
  // You can then refer to these in the twig template like
  // {{ page.title }}, {{ env }} and {{ config.domain }}
})
app.listen(3000)
```

**Comment by gageleb** [comment]
Yeah, I'm using express now, and that's what I'm using, but in PHP's twig, you set globals in a config before the rendering process. So it would be something like 

```
app.set('twig globals', {
    title: 'page title'
});
```

and then it could be referenced within the template without setting it in the object that res.render uses.

**Comment by ldonis** [comment]
This is the only way I can make it work, closely to the PHP implementation
```js
// twigjs setup
twig.extendFunction('template', () => { 
    return new templateHelper() 
})

// twigjs view
{{ template().methodOfClass('text') }}
```
I'm using this workaround until found a better way to implement something like addGlobals

**Comment by talyguryn** [comment]
You can create a middleware and define global vars

```js
/**
 * Set global vars for Twig tempates
 */
app.use((req, res, next) => {
  res.locals = {
    HOSTNAME: `${req.protocol}://${req.get('host')}`,
    TITLE: 'My amazing site'
  };
  next();
});
```

**Comment by NotNotMarshall** [comment]
> Yeah, I'm using express now, and that's what I'm using, but in PHP's twig, you set globals in a config before the rendering process. So it would be something like
> 
> ```
> app.set('twig globals', {
>     title: 'page title'
> });
> ```
> 
> and then it could be referenced within the template without setting it in the object that res.render uses.

I love you!
